src/controller/datacontroller.js

```diff
@@ -306,6 +336,49 @@ export default class DataController {
 
 mix( DataController, EmitterMixin );
 
+// Traverses given DocumentFragment and searches elements which marks marker range. Found element is removed from
+// DocumentFragment but path of this element is stored in a Map which is then returned.
+//
+// @param {module:engine/view/documentfragment~DocumentFragment} documentFragment Model DocumentFragment.
+// @returns {Map} Map with markers data.
+function extractMarkersFromModelFragment( documentFragment ) {
+	const markerStamps = new Set();
+
+	// Create ModelTreeWalker.
+	const walker = new ModelTreeWalker( {
+		startPosition: ModelPosition.createAt( documentFragment, 0 ),
+		ignoreElementEnd: true,
+		shallow: false
+	} );
+
+	// Walk through DocumentFragment and collect marker elements.
+	for ( const value of walker ) {
+		// Check if current element is a marker stamp.
+		if ( value.item.name == '$marker' ) {
+			markerStamps.add( value.item );
+		}
+	}
+
+	// Walk through collected marker elements store its path and remove its from the DocumentFragment.
+	return Array.from( markerStamps ).reduce( ( result, stamp ) => {
```

Comment:
Loops FTW!

On serious note, this looks overcomplicated. Simple loop over `markerStamps` would be okay too.